scapy/layers/ipsec.py

```diff
@@ -887,11 +1168,20 @@ def check_spi(self, pkt):
 
     def _encrypt_esp(self, pkt, seq_num=None, iv=None, esn_en=None, esn=None):
 
-        if iv is None:
-            iv = self.crypt_algo.generate_iv()
+        # Here we use either aead algo or crypt + auth
+        if self.aead_algo is None:
+            if iv is None:
+                iv = self.crypt_algo.generate_iv()
+            else:
+                if len(iv) != self.crypt_algo.iv_size:
+                    raise TypeError('iv length must be %s' % self.crypt_algo.iv_size)  # noqa: E501
         else:
-            if len(iv) != self.crypt_algo.iv_size:
-                raise TypeError('iv length must be %s' % self.crypt_algo.iv_size)  # noqa: E501
+            if iv is None:
+                iv = self.aead_algo.generate_iv()
```

Comment:
Typically if `crypt_algo = aead_algo` this wouldn't be needed.

Reply:
Hello.

Again, crypt algo is not equivalent to AEAD algo.

In addition, crypt algo implementation in cryptography library is not equivalent to AEAD implementation aswell.

For example, aead algo does not have "mode" parameter.

Reply:
AEAD algo are not equivalent to Crypt algos. See CHACHA20_POLY1305 comment. RFC detail explanation here: https://datatracker.ietf.org/doc/html/rfc7539